14.0 dev training - not to be merged (#60033)
odoo/odoo: doc/howtos/rdtraining/E_unittest.rst

.. _howto/rdtraining/E_unittest:

=============================
Advanced E: Python Unit Tests
=============================

.. tip:: This tutorial assumes you followed the Core Training.

  To do the exercise, fetch the branch 14.0-core from the repository XXX.
  It contains a basic module we will use as a starting point

**Reference**:
`Odoo's Test Framework: Learn Best Practices <https://www.youtube.com/watch?v=JEIscps0OOQ>`__
(Odoo Experience 2020) on Youtube.

Writing tests is a necessity for multiple reasons. Here is a non exhaustive list:

* Ensure it will not be broken in the future
* Define the scope of your code
* Give examples of use cases
* It is one way to technically document the code
* Help you develop by defining your goal before working towards it

> 

Reviewer: I think before mentioning integration bots, it could be interesting to explain how a test can be run locally.

Integration Bots
================

.. note:: This section is only for Odoo employees and people that are contributing to
  `github.com/odoo`. We highly recommend having your own CI if it is not the case.

> The tests would be useless without a Continuous Integration (CI). This is why we have some bots

Reviewer: ```suggestion
When a test is written, it is important to make sure it always passes when modifications are applied to the source code. To automatize this task, we use a development practice called Continuous Integration (CI). This is why we have some bots
```

running all the tests at different moments. Whether you are working at Odoo or not, if you are
trying to merge something inside `odoo/odoo`, `odoo/enterprise`, `odoo/upgrade` or on odoo.sh, you
will have to go through the CI. If you are working on another project, you should think of adding
your own CI.

Runbot
------

**Reference**: the documentation related to this topic can be found in
`Runbot FAQ <https://runbot.odoo.com/doc>`__.

Most of the tests are run on `Runbot <https://runbot.odoo.com>`__ every time a commit is pushed on
GitHub.

You can see the state of a commit/branch by filtering on the runbot dashboard.

A **bundle** is created for each branch. A bundle consists of a configuration and contains the
batches.

A **batch** is a set of builds, depending on the parameters of the bundle.

> A batch is greened if all the builds are greened.

Reviewer: ```suggestion
A batch is green (i.e. passes the tests) if all the builds are green.
```

A **build** is when we launch a server. It can be divided in sub-builds. Usually there are builds
for the community version, the enterprise version (only if there is an enterprise branch but you
can force the build), and the migration of the branch.

> A build is greened if every sub-build is greened.

Reviewer: ```suggestion
A build is green if every sub-build is green.
```

A **sub-build** only does some parts of what a full build does. It is used to speed up the CI
process. Generally it is used to split the post install tests in 4 parallel instances.

> A sub-build is greened if all the tests are passing and there are no errors/warnings logged.

Reviewer: ```suggestion
A sub-build is green if all the tests are passing and there are no errors/warnings logged.
```

.. note::

>   * Robodoo doesn't care what the changes were. It will always run the same tests, meaning it

Reviewer: Maybe early to talk about robodoo.

```suggestion
  * All tests are run whatever the modifications done. Correcting a typo in an error message or refactoring a whole module triggers the same tests. It
```

>     will install all the modules. This means something might not work if runbot greened but your

Reviewer: ```suggestion
    will install all the modules. This means something might not work if runbot is green but your
```

    changes depend on something you don't depend on.

>   * The localization modules are not installed on runbot (except the generic one), some modules

Reviewer: ```suggestion
  * The localization modules (i.e. country-specific modules) are not installed on runbot (except the generic one), some modules
```

    with external dependencies can be excluded also.
  * There is a nightly build running additional tests, like module operations, localization, single
    module installs, multi-builds for nondeterministic bugs, etc.
    These are not kept in the standard CI to shorten the time of execution.

You can also login on a build built by runbot. There are 3 users usable: `admin`, `demo` and
`portal`. The password is the same as the login. This is useful to quickly test things on different
versions without having to build it locally. The full logs are also available; these are used for
monitoring.

Robodoo
-------

You will most likely have to gain a little bit more experience before having the rights to summon
robodoo, but here are a few remarks anyways.

Robodoo is the guy spamming the CI status as tags on your PRs, but he is also the guy that kindly
integrates your commits on the main repositories.

> When the last batch is greened, the reviewer can ask robodoo to merge your PR (actually it is more

Reviewer: ```suggestion
When the last batch is green, the reviewer can ask robodoo to merge your PR (actually it is more
```

a `rebase` than a `merge`). It will then go to the mergebot.

Mergebot
--------

`Mergebot <https://mergebot.odoo.com>`__ is the last testing phase before merging a PR.

It will take the commits in your branch not yet present on the target, stage it and rerun the tests
one more time, including the enterprise version even if you are only changing something in
community.

> 

Reviewer: Maybe add some info about "Staging failed" error message?

More generally, there could be a part explaining what to do if the runbot is red. Check the logs, run the tests locally (maybe mention `--test-tags` and `--test-file`).

Modules
=======

Because Odoo is modular, the tests need to be modular also. This means the modules are defined in
the module that adds the functionality you add; and that tests cannot depend on functionality
coming from modules your module doesn't depend on.

If the behavior you want to test can be changed by the installation of another module, you need to

> ensure that the tag `at_install` is set; otherwise you can use the tag `post_install` to speed up

Reviewer: Maybe add a code snippet to clarify?

the CI, and ensure it is not changed if it shouldn't.

Writing a test
==============

> **Reference**: the documentation related to this topic can be found in
> `Python unittest <https://docs.python.org/3/library/unittest.html>`__.
> 
> **Reference**: the documentation related to this topic can be found in
> :ref:`Testing Odoo<reference/testing>`.

Reviewer: ```suggestion
**Reference**: the documentation related to this topic can be found in
`Python unittest <https://docs.python.org/3/library/unittest.html>`__ and :ref:`Testing Odoo<reference/testing>`.
```

Here are a few things to take into consideration before writing a test

> * The tests should be independent from the data currently in the database

Reviewer: ```suggestion
* The tests should be independent from the data currently in the database (including demo data)
```

* Tests should not impact the database by leaving/changing residual data. This is usually done by

>   the test framework

Reviewer: ```suggestion
  the test framework. You must never call `cr.commit` in a test.
```

Author: This is not proper to tests, is it? Calling `cr.commit` will have undesired effects almost anywhere, shouldn't it be part of the guidelines?

Reviewer: indeed, it should be called only in exceptional cases.

  in python. This also applies in the tests. If you want to emulate a Form view, you can use
  ``odoo.tests.common.Form``.

The tests should be located in a ``tests`` folder in the root of your module. Each test file name
should start with `test_` and be imported in the ``__init__.py`` of the test folder. You shouldn't
import the test folder/module in the ``__init__.py`` of the module.

.. code-block:: bash

  estate
  ├── models
  │   ├── *.py
  │   └── __init__.py
  ├── tests
  │   ├── test_*.py
  │   └── __init__.py
  ├── __init__.py
  └── __manifest__.py

.. note:: Some older tests are extending ``odoo.tests.common.TransactionCase``, but they are less
  scalable. The difference is that the setup is done per test method and not per test class.
  The data changed are rollbacked between each test in `SavepointCase` to have the same behavior as
  in `TransactionCase`.

All the tests should extend ``odoo.tests.common.SavepointCase``. You usually define a
``setUpClass``, and the tests. After doing the `setUpClass`, you have an `env` available on the
class and can start interacting with the ORM.

> These test classes are build on top of the ``unittest`` module.

Reviewer: ```suggestion
These test classes are built on top of the ``unittest`` module.
```
